### PR TITLE
Deploy referenced views to stage

### PR DESCRIPTION
**Pull request by scholtzan**
Related to https://app.circleci.com/pipelines/github/mozilla/bigquery-etl/33384/workflows/b9a2207f-5d42-45fb-b2ac-2616967916bc/jobs/369291

The stage deploys should deploy views that are referenced in queries to stage. We are already doing this for views referenced in other views. Currently, we are missing cases where a query is referencing a view, which might not have changed but is referencing another view with changes. The query still references the original deployed in production in those cases. 
The changes made in this PR should fix this.

Checklist for reviewer:

- [ ] Commits should reference a bug or github issue, if relevant (if a bug is referenced, the pull request should include the bug number in the title).
- [ ] If the PR comes from a fork, trigger integration CI tests by running the [Push to upstream workflow](https://github.com/mozilla/bigquery-etl/actions/workflows/push-to-upstream.yml) and provide the `<username>:<branch>` of the fork as parameter. The parameter will also show up
in the logs of the `manual-trigger-required-for-fork` CI task together with more detailed instructions.
- [ ] If adding a new field to a query, ensure that the schema and dependent downstream schemas have been updated.
- [ ] When adding a new derived dataset, ensure that data is not available already (fully or partially) and recommend extending an existing dataset in favor of creating new ones. Data can be available in the [bigquery-etl repository](https://github.com/mozilla/bigquery-etl), [looker-hub](https://github.com/mozilla/looker-hub) or in [looker-spoke-default](https://github.com/mozilla/looker-spoke-default/tree/e1315853507fc1ac6e78d252d53dc8df5f5f322b).

For modifications to schemas in restricted namespaces (see [`CODEOWNERS`](https://github.com/mozilla/bigquery-etl/blob/main/CODEOWNERS)):
- [ ] Follow the [change control procedure](https://docs.google.com/document/d/1TTJi4ht7NuzX6BPG_KTr6omaZg70cEpxe9xlpfnHj9k/edit#heading=h.ttegrcfy18ck)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/DENG-3625)

### CODE REVIEW

File: bigquery_etl/cli/stage.py [hunk 1]
```diff
@@ -260,13 +261,35 @@ def _udf_dependencies(artifact_files):
...
+            # check query files for view dependencies
+            # views need to be deployed to stage as they might reference other changed artifacts upstream
+            # without getting changed themselves
+            dependencies = extract_table_references(
+                render(dep_file.name, dep_file.parent)
+            )
+            for dependency in dependencies:
+                dependency_components = dependency.split(".")
+                if len(dependency_components) == 3:
+                    # views are always fully qualified, so they'll consist of 3 parts
```
Comment (sean-rose) [inline bigquery_etl/cli/stage.py:282]: I don't think view references are always fully qualified ([example](https://github.com/mozilla/bigquery-etl/blob/dd824d66ca2f8a1ca3f546cab758f041d7794155/sql/moz-fx-data-shared-prod/fenix_derived/events_daily_v1/query.sql#L32)).

Comment (kik-kik) [reply bigquery_etl/cli/stage.py:282]: That is correct, I believe only inside `view.sql` we are required to use fully qualified.

File: bigquery_etl/cli/stage.py [hunk 2]
```diff
@@ -260,13 +261,51 @@ def _udf_dependencies(artifact_files):
...
+                    [
+                        depdendency_project_id,
+                        dependency_dataset_id,
+                        dependency_table_id,
+                    ] = dependency_components
+                # Fully qualify the reference:
+                elif len(dependency_components) == 2:
+                    depdendency_project_id = dep_file.parent.parent.parent.name
+                    [dependency_dataset_id, dependency_table_id] = dependency_components
+                elif len(dependency_components) == 1:
```
Comment (kik-kik) [inline bigquery_etl/cli/stage.py:292]: If there is only one component wouldn't it be unlikely the reference contained a `.` character?

File: bigquery_etl/cli/stage.py [hunk 3]
```diff
@@ -260,13 +261,51 @@ def _udf_dependencies(artifact_files):
...
+
+                file_path = (
+                    Path(sql_dir)
+                    / depdendency_project_id
+                    / dependency_dataset_id
+                    / dependency_table_id
+                    / VIEW_FILE
+                )
+                if file_path.exists():
+                    view_dependency_files.append(file_path)
```
Comment (kik-kik) [inline bigquery_etl/cli/stage.py:307]: So is the idea here that we recurse over query files?

File: bigquery_etl/cli/stage.py [hunk 4]
```diff
@@ -260,13 +261,51 @@ def _udf_dependencies(artifact_files):
...
+            # check query files for view dependencies
+            # views need to be deployed to stage as they might reference other changed artifacts upstream
+            # without getting changed themselves
+            dependencies = extract_table_references(
+                render(dep_file.name, dep_file.parent)
+            )
+            for dependency in dependencies:
+                dependency_components = dependency.split(".")
+
+                if len(dependency_components) == 3:
```
Comment (kik-kik) [inline bigquery_etl/cli/stage.py:282]: Could this be a good place to use `match`?
https://docs.python.org/3/tutorial/controlflow.html#match-statements

File: bigquery_etl/cli/stage.py [hunk 5]
```diff
@@ -260,13 +261,51 @@ def _udf_dependencies(artifact_files):
...
+                    [
+                        depdendency_project_id,
+                        dependency_dataset_id,
+                        dependency_table_id,
+                    ] = dependency_components
+                # Fully qualify the reference:
+                elif len(dependency_components) == 2:
+                    depdendency_project_id = dep_file.parent.parent.parent.name
+                    [dependency_dataset_id, dependency_table_id] = dependency_components
```
Comment (kik-kik) [inline bigquery_etl/cli/stage.py:291]: I don't think we need the `[]` here to unpack the values?

**Review by kik-kik — APPROVED** [review]
r+wc

File: bigquery_etl/cli/stage.py (hunk 3 above)
Comment (scholtzan) [reply bigquery_etl/cli/stage.py:307]: yup

File: bigquery_etl/cli/stage.py [hunk 6]
```diff
@@ -260,13 +261,57 @@ def _udf_dependencies(artifact_files):
...
+                    None,
+                    None,
+                    None,
+                )
+
+                match len(dependency_components):
+                    case 3:
+                        dependency_project_id,
+                        dependency_dataset_id,
+                        dependency_table_id = dependency_components
```
Comment (sean-rose) [inline bigquery_etl/cli/stage.py:291]: I was surprised this is valid Python syntax.  I know @kik-kik suggested not using list/tuple syntax for one of the unpacking assignment statements, but I think that was intended just for that case where it all fit on one line.  With Black auto-formatting this so poorly (IMO) I think using list/tuple syntax in this case would be more readable.

File: bigquery_etl/cli/stage.py [hunk 7]
```diff
@@ -260,13 +261,57 @@ def _udf_dependencies(artifact_files):
...
+                        dependency_project_id = dep_file.parent.parent.parent.name
+                        dependency_dataset_id, dependency_table_id = (
+                            dependency_components
+                        )
+                    case 1:
+                        dependency_project_id = dep_file.parent.parent.parent.name
+                        dependency_dataset_id = dep_file.parent.parent.name
+                        dependency_table_id = dependency_components[0]
+                    case _:
+                        return
```
Comment (sean-rose) [inline bigquery_etl/cli/stage.py:303]: I don't think we'd want to stop and return nothing at all, just skip this particular dependency.
```suggestion
                        continue
```

**Pull request merged**